src/main/java/nextstep/auth/authentication/AuthenticationInterceptor.java

```diff
@@ -0,0 +1,45 @@
+package nextstep.auth.authentication;
+
+import java.io.IOException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import nextstep.auth.application.UserDetailService;
+import nextstep.auth.authentication.converter.AuthenticationConverter;
+import nextstep.auth.context.Authentication;
+import nextstep.auth.context.SecurityContextHolder;
+import nextstep.auth.domain.UserDetails;
+import org.springframework.web.servlet.HandlerInterceptor;
+
+public abstract class AuthenticationInterceptor implements HandlerInterceptor {
+
+    private final UserDetailService userDetailService;
+    private final AuthenticationConverter authenticationConverter;
+
+    public AuthenticationInterceptor(UserDetailService userDetailService,
+            AuthenticationConverter authenticationConverter) {
+        this.userDetailService = userDetailService;
+        this.authenticationConverter = authenticationConverter;
+    }
+
+    @Override
+    public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler) throws Exception {
+        AuthenticationToken token = authenticationConverter.convert(request);
+        Authentication authenticate = authenticate(token);
+        afterAuthentication(request, response, authenticate);
+        return false;
+    }
+
+    public Authentication authenticate(AuthenticationToken token) {
+        String principal = token.getPrincipal();
+        UserDetails userDetails = userDetailService.loadUserByUsername(principal);
+        userDetails.checkPassword(token.getCredentials());
+
+        return new Authentication(userDetails);
```

Comment:
@hiro032 
만약에 LoginMember와 UserDetails를 분리를 한다고 하면,
LoginMember는 UserDetails를 상속받지 않은 단순 DTO의 역할을 하고
UserDetails를 상속받은 custom UserDetails 클래스를 새로 만든다고 가정했을 때,

기존 아키텍처에서 크게 변경을 하지 않으려면 아래와 같이 코드를 짤 수 있을 것 같아요!

```suggestion
        String principal = token.getPrincipal();
        UserDetails userDetails = userDetailService.loadUserByUsername(principal);
        userDetails.checkPassword(token.getCredentials());

        return new Authentication(userDetails.toLoginMember()); // `toLoginMember`는 LoginMember 객체를 리턴
```

코멘트 드린 부분이긴 하지만 미션과는 큰 연관성은 없기 때문에 꼭 할 필요는 없습니다.
민봉님 판단에 맡길게요 :)

Reply:
이해했습니다 감사합니다! 😊